earth_enterprise/src/common/SharedString.h

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright 2019 The Open GEE Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef GEO_EARTH_ENTERPRISE_SRC_COMMON_SHAREDSTRING_H_
+#define GEO_EARTH_ENTERPRISE_SRC_COMMON_SHAREDSTRING_H_
+
+#include <string>
+#include <unordered_map>
+#include <vector>
+#include <iostream>
+#include <stdint.h>
+#include <assert.h>
+#include <common/notify.h>
+
+// The goal of this class is to keep in memory only one copy of the name of
+// a cached asset or asset version. Previously, each asset stored the full names
+// of all of its children, parents, listeners, and inputs, resulting in tens of
+// thousands of duplicates of each string in very large projects.
+//
+// With this class, only an integer identifier is stored in each instance (thus
+// drastically reducing the amount of memory needed in large projects), and
+// that identifier is used to retrieve the string value when it is requested.
+// Since the string-to-identifier ratio is 1:1, equality comparisons are also 
+// much faster as we only need to compare the identifiers.
+class SharedString {
+protected:
+   class StringStorage {
+      private:
+        std::unordered_map<uint32_t, std::string> refFromKeyTable;
+        std::unordered_map<std::string, uint32_t> keyFromRefTable;
+        static uint32_t nextID;

```

Comment:
Just thinking, would it be better to use a hash instead of an ID? That way, you could actually check to see if the string is already in the lookup table without doing a search. I guess it depends on the performance of the hash algorithm as to which would work better.

Reply:
Since `unordered_map` uses hashing internally, I'm not sure how much improvement that would buy when looking things up.